src/ir_generator.rs

```diff
@@ -0,0 +1,193 @@
+use crate::compiler::Compiler;
+use crate::errors::{Severity, SourceError};
+use crate::parser::{AstNode, NodeId};
+use nu_protocol::ast::{Math, Operator};
+use nu_protocol::ir::{Instruction, IrBlock, Literal};
+use nu_protocol::{RegId, Span};
+
+/// Generates IR (Intermediate Representation) from nu AST.
+pub struct IrGenerator<'a> {
+    // Immutable reference to a compiler after the typechecker pass
+    compiler: &'a Compiler,
+    errors: Vec<SourceError>,
+
+    instructions: Vec<Instruction>,
+    register_count: u32,
+    file_count: u32,
+}
+
+impl<'a> IrGenerator<'a> {
+    pub fn new(compiler: &'a Compiler) -> Self {
+        Self {
+            compiler,
+            errors: Default::default(),
+            instructions: Default::default(),
+            register_count: 0,
+            file_count: 0,
+        }
+    }
+
+    /// Generates the IR from the given state of the compiler.
+    /// After this is called, use `block` and `errors` to get the result.
+    pub fn generate(&mut self) {
+        if self.compiler.ast_nodes.is_empty() {
+            return;
+        }
+        let Some(reg) = self.generate_node(NodeId(self.compiler.ast_nodes.len() - 1)) else {
+            return;
+        };
+        self.instructions.push(Instruction::Return { src: reg });
+    }
+
+    /// Returns generated IR block.
+    ///
+    /// Call `generate` before using this method and ensure there are no errors.
+    pub fn block(self) -> IrBlock {
+        // TODO: properly generate the spans
+        // TODO: figure out what to do with AST, as this parser has different
+        // representation of AST than the old parser.
```

Comment:
The references to the AST in IR are only due to some parts of IR still having dependencies on the AST evaluator, so I think the best course of action would be to ignore the AST in the new parser and work on completely removing the AST evaluator in Nushell.

Reply:
I fully agree, I will look into the nushell to see if we can somehow completely eliminate the dependency on AST. I believe I saw some DebugContext that was accepting the AST node, but I need to dig into the code a little more.